### PR TITLE
feat(Redis Memory storage): Caching module with redis and memory stor…

### PR DESCRIPTION
**Pull request by chandgi**
BREAKING CHANGE: Simple and extensible caching module with redis and memory storage and supporting decorators.

This caches function response using the given options. Works with different strategies and storages. Uses all arguments to build an unique key.

### CODE REVIEW

File: lib/cache/class/index.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    catalog = null,
+    catalogDuration = -1,
+    id,
+    duration = 60,
+    queryMethod,
+    queryArgs = [null],
+    callback,
+  ) {
+    let self = this;
+    let dataMethod = q.async(function*() {
```
Comment (JamilOmar) [inline lib/cache/class/index.js]: let's user await and async instead of q

File: lib/cache/decorators/cache.clear.decorator.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+          return descriptor.value!.apply(this, args);
+        }
+      }
+
+      const entry = await cachingStrategy.getItem(hashKey);
+
+      if (entry) {
+        return await cachingStrategy.deleteItem(hashKey);
+      } else {
+        return undefined;
```
Comment (JamilOmar) [inline lib/cache/decorators/cache.clear.decorator.ts]: only return

File: lib/cache/decorators/cache.decorator.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    methodName: string,
+    descriptor: PropertyDescriptor,
+  ) {
+    const originalMethod = descriptor.value;
+    const className = target.constructor.name;
+    debugger;
+
+    descriptor.value = async function(...args: any[]) {
+      const generatedCacheKey =
+        !Array.isArray(args) || !args.length
```
Comment (JamilOmar) [inline lib/cache/decorators/cache.decorator.ts]: if is not array it shouldnt have length

File: lib/cache/decorators/cache.decorator.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+          ? `${className}:${methodName}`
+          : `${className}:${methodName}:${JSON.stringify(args)}`;
+      const cacheKey = !_.isEmpty(options.cacheKey)
+        ? options.cacheKey
+        : generatedCacheKey;
+      let cachingStrategy = _.get(global, 'LABSHARE_CACHE', undefined);
```
Comment (JamilOmar) [inline lib/cache/decorators/cache.decorator.ts]: use constants

File: lib/cache/decorators/cache.decorator.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+      if (cachingStrategy) {
+        if (options && options.noop) {
+          return originalMethod.apply(this, args);
+        }
+      }
+      if (_.isUndefined(cachingStrategy)) {
+        return undefined;
```
Comment (JamilOmar) [inline lib/cache/decorators/cache.decorator.ts]: only return

File: lib/cache/middleware/base/index.js [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+ * @Throws {Error} if is unable to connect with redis.
+ * @param {Object} redis - Redis API configuration.
+ * @param {Number} maxTime - The cache maxTime.
+ * @param {String} prefix - Cache prefix for identification. Default : '' .
+ * @param {String} catalog - Default catalog. Default : null .
+ * @param {Number} catalogDuration - Catalog duration time. Default : -1 (not expire)'' .
+ * {Object} logger - Error logging provider. It must define an `error` function. Default: null
+ * @constructor
+ */
+class Middleware {
```
Comment (JamilOmar) [inline lib/cache/middleware/base/index.js]: this is not required

File: lib/cache/providers/cache.strategy.provider.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import * as _ from 'lodash';
+
+export class CacheStrategyResolverProvider
+  implements Provider<ExpirationStrategy> {
+  constructor(
+    @inject(ConfigBindings.CONFIG, {optional: true})
+    protected labShareConfiguration: ApplicationConfig,
+  ) {}
+
+  async value(): Promise<ExpirationStrategy> {
```
Comment (JamilOmar) [inline lib/cache/providers/cache.strategy.provider.ts]: change ExpirationStrategy to LabShareCache

File: lib/cache/providers/cache.strategy.provider.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    let provider;
+    if (
+      _.get(this.labShareConfiguration, 'cache.strategy', 'memory') === 'redis'
+    ) {
+      const redisOptions = _.get(
+        this.labShareConfiguration,
+        'cache.redisOptions',
+      );
+      provider = new ExpirationStrategy(new RedisStorage(redisOptions));
+      _.set(global, 'LABSHARE_CACHE', provider);
```
Comment (JamilOmar) [inline lib/cache/providers/cache.strategy.provider.ts]: Use constants

File: lib/cache/storages/redis.storage.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        });
+        this.client.on('end', () => {
+          this.connectionStatus = redisStatus.DISCONNECTED;
+        });
+        this.client.on('error', () => {
+          this.connectionStatus = redisStatus.DISCONNECTED;
+        });
+      }
+    } catch (error) {
+      console.log('connection error', error);
```
Comment (JamilOmar) [inline lib/cache/storages/redis.storage.ts]: throw the error

File: lib/cache/storages/redis.storage.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    }
+  }
+
+  public async getItem<T>(key: string | undefined): Promise<T> {
+    if (
+      this.failsafeMode &&
+      this.connectionStatus === redisStatus.DISCONNECTED
+    ) {
+      return undefined;
```
Comment (JamilOmar) [inline lib/cache/storages/redis.storage.ts]: only return

File: lib/cache/storages/redis.storage.ts [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  public async setItem(key: string, content: any): Promise<void> {
+    if (
+      this.failsafeMode &&
+      this.connectionStatus === redisStatus.DISCONNECTED
+    ) {
+      return undefined;
+    }
+    if (typeof content === 'object') {
```
Comment (JamilOmar) [inline lib/cache/storages/redis.storage.ts]: you can user lodash _.isobject

File: lib/cache/strategies/expiration.strategy.ts [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+import {IOptions, IExpiringCacheItem} from '../types';
+
+export class ExpirationStrategy extends AbstractBaseStrategy {
+  constructor(storage: IStorage) {
+    super(storage);
+  }
+
+  public async getItem<T>(key: string): Promise<T> {
+    const item = await this.storage.getItem<IExpiringCacheItem>(key);
+    if (item && item.meta && item.meta.ttl && this.isItemExpired(item)) {
```
Comment (JamilOmar) [inline lib/cache/strategies/expiration.strategy.ts]: check lodash method _.has

File: lib/cache/strategies/expiration.strategy.ts [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+      await this.storage.setItem(key, undefined);
+      return undefined;
```
Comment (JamilOmar) [inline lib/cache/strategies/expiration.strategy.ts]: only use return

File: lib/cache/strategies/expiration.strategy.ts [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+    content: any,
+    options: IOptions,
+  ): Promise<void> {
+    options = {ttl: 60, isLazy: true, isCachedForever: false, ...options};
+
+    let meta = {};
+
+    if (!options.isCachedForever) {
+      meta = {
+        ttl: options.ttl * 1000,
```
Comment (JamilOmar) [inline lib/cache/strategies/expiration.strategy.ts]: no magic numbers, change 1000 to a constant

File: lib/cache/types.ts [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+  noop?: boolean;
+  refreshCache?: boolean;
+}
+
+import {Request, OperationArgs} from '@loopback/rest';
+
+/**
+ * A function to perform REST req/res logging action
+ */
+export interface LogFn {
```
Comment (JamilOmar) [inline lib/cache/types.ts]: This is not required

File: lib/cache/component.ts [hunk 16]
```diff
@@ -0,0 +1,13 @@
+import { Component, ProviderMap } from '@loopback/core';
+import { CacheBindings } from './keys';
+import { CacheStrategyResolverProvider, TimerProvider } from './providers';
+/**
+ * Implements a configuration component.
+ * @access public
+ */
+export class LbServicesCacheComponent implements Component {
+  providers?: ProviderMap = {
+    [CacheBindings.TIMER.key]: TimerProvider,
```
[the remark this answers is not in the document]
Comment (chandgi) [reply lib/cache/component.ts]: These are outdated file, checked in clean version, please check the latest checked in files.

File: lib/cache/decorators/cache.clear.decorator.ts [hunk 17]
```diff
@@ -0,0 +1,32 @@
+import * as _ from 'lodash';
+
+export function CacheClear(options: any): Function {
+    return function (target: any, methodName: string, descriptor: PropertyDescriptor) {
+        const className = target.constructor.name;
+       
+        descriptor.value = async function (...args: any[]) {
+            const provider = _.get(global, 'LABSHARE_CACHE', undefined);
```
[the remark this answers is not in the document]
Comment (chandgi) [reply lib/cache/decorators/cache.clear.decorator.ts]: Changed and moved to constant file

File: lib/cache/decorators/cache.clear.decorator.ts [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+            let cachingStrategy = provider;
```
[the remark this answers is not in the document]
Comment (chandgi) [reply lib/cache/decorators/cache.clear.decorator.ts]: changed to Const instead of the let

File: lib/cache/middleware/base/index.js (hunk 6 above)
Comment (chandgi) [reply lib/cache/middleware/base/index.js]: These are existing old code, removing it.

File: lib/cache/types.ts (hunk 15 above)
Comment (chandgi) [reply lib/cache/types.ts]: Removed from code base.

File: lib/cache/strategies/expiration.strategy.ts (hunk 14 above)
Comment (chandgi) [reply lib/cache/strategies/expiration.strategy.ts]: added constants file

**Pull request merged**